### DISCUSSION TITLE
Noob help with Keys/useListData in a Select Component

### DISCUSSION THREAD

**Discussion by NickWoodward** [adobe/react-spectrum#6620] [Q&A]
Hi there,

Sorry if this isn't the place for this sort of beginner question, but I'm really struggling to make a reusable Select Component with Keys and/or useListData. 

So far I can make it sort of work with strings and `useState`, but it means I've only got access to the selected index:

Live Demo [here](https://codesandbox.io/p/devbox/wonderful-burnell-forked-srfw65?file=%2Fsrc%2FJobTypeSelect.tsx%3A12%2C20&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxuq704500772a699proyjsa%2522%252C%2522sizes%2522%253A%255B71.7605633802817%252C28.239436619718305%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxuq704500742a69dqn6bg57%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxuq704500762a69yoj1jd12%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxuq704500752a69g5kw4ymx%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B66.8294350857958%252C33.170564914204206%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxuq704500742a69dqn6bg57%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxtg87yn00012a69yrf5ij8d%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxtih51d00022a699d949l78%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A49%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A49%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FItemSelect.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxubupkx00022a69vfg7od89%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A1%252C%2522startColumn%2522%253A23%252C%2522endLineNumber%2522%253A1%252C%2522endColumn%2522%253A23%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxujvugb00022a69q9cbq3ym%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A14%252C%2522startColumn%2522%253A29%252C%2522endLineNumber%2522%253A14%252C%2522endColumn%2522%253A29%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FWageSlider.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxv561hk00022a694mefkgum%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A12%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A12%252C%2522endColumn%2522%253A20%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FJobTypeSelect.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxuq704500742a69dqn6bg57%2522%252C%2522activeTabId%2522%253A%2522clxv561hk00022a694mefkgum%2522%257D%252C%2522clxuq704500752a69g5kw4ymx%2522%253A%257B%2522id%2522%253A%2522clxuq704500752a69g5kw4ymx%2522%252C%2522activeTabId%2522%253A%2522clxuqqd7j00kl2a69kgk1lpgd%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A5173%252C%2522id%2522%253A%2522clxuqqd7j00kl2a69kgk1lpgd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clxuq704500762a69yoj1jd12%2522%253A%257B%2522id%2522%253A%2522clxuq704500762a69yoj1jd12%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxtg87yn00032a69gisj4210%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clxtg86fr000hd9f89rtlhhys%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clxtg8gk500652a69yv83z3oe%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clxtg87yn00032a69gisj4210%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A16.785585136866658%257D)

Code Snippets:
```
export default function App() {
  const [tags, setTags] = useState<string[]>([]);
    const items = [
      { id: 1, name: "Permanent" },
      { id: 2, name: "Temporary" },
      { id: 3, name: "Contract" },
    ];
    const addTag = (item: string) => {
      if (tags.includes(item.toString())) return;
      setTags([...tags, item]);
    };
    const removeTag = (item: string) => {
      setTags((tags) => tags.filter((tag) => tag !== item));
    };
    
    return(
      //...
      <JobTypeSelect items={items} setJobType={addTag} className="" />
        {tags.map((tag) => (
          <Button onPressStart={() => removeTag(tag)}>
            <span>{tag}</span>
            <CloseIcon  />
          </Button>
        ))}
    )
```

// JobTypeSelect.tsx
```
interface Props {
  className: string;
  setJobType: (item: string) => void;
  items: { id: number; name: string }[];
}
export const JobTypeSelect = ({ className, items, setJobType }: Props) => {
  return (
    <ItemSelect
      label="Job Type"
      items={items}
      onSelectionChange={(item) => {
        return setJobType(item.toString());
      }}
    >
      {(item) => <ListBoxItem key={item.id}>{item.name}</ListBoxItem>}
    </ItemSelect>
  );
};
```

Obviously this isn't the right way to do it, but every time I try and replace my state with `useListData` I struggle typing the `onSelectionChange` and other properties I try and pass, even with the api [here](https://react-spectrum.adobe.com/react-stately/useListData.html)

Really would appreciate any guidance, or even links to example Components that are both typed and more reusable (most examples I've found tend to use hooks)

Thanks a lot!

Nick

**Comment by snowystinger** [comment]
Your codesandbox is private, you'll need to mark it as public. 
We have some examples here using useListData in RAC https://react-spectrum.adobe.com/react-aria/ListBox.html#reorderable

I think you're getting mixed up between item vs the selection key "id". It may help to log out `item.toString` in your onSelectionChange.

**Comment by NickWoodward** [comment]
*edit: I think I fixed it 😊

Hi @snowystinger, thanks for the reply,

So I decided to ditch using my own state and try to implement `useListData`

Have spent a long time on this so I think I'm close, and I hope it's made for a better question. I'm just struggling to type the `list` as I pass it to my `ItemSelect` component. I get the following error:

> Type 'ListData<T>' is not assignable to type 'ListData<{ id: number; name: string; }>'.
> Type 'T' is not assignable to type '{ id: number; name: string; }'.typescript(2322)

My app structure is like this:
Codepen Demo [here](https://codesandbox.io/p/devbox/test-28j3m5?file=%2Fsrc%2FApp.tsx%3A22%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxuq704500772a699proyjsa%2522%252C%2522sizes%2522%253A%255B71.7605633802817%252C28.239436619718305%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxuq704500742a69dqn6bg57%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxuq704500762a69yoj1jd12%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxuq704500752a69g5kw4ymx%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B88.33751667986571%252C11.662483320134285%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxuq704500742a69dqn6bg57%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxtg87yn00012a69yrf5ij8d%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxtih51d00022a699d949l78%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A49%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A49%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FItemSelect.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxubupkx00022a69vfg7od89%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A22%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A22%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxujvugb00022a69q9cbq3ym%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A14%252C%2522startColumn%2522%253A29%252C%2522endLineNumber%2522%253A14%252C%2522endColumn%2522%253A29%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FWageSlider.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clxv561hk00022a694mefkgum%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A12%252C%2522startColumn%2522%253A20%252C%2522endLineNumber%2522%253A12%252C%2522endColumn%2522%253A20%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FJobTypeSelect.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxuq704500742a69dqn6bg57%2522%252C%2522activeTabId%2522%253A%2522clxubupkx00022a69vfg7od89%2522%257D%252C%2522clxuq704500752a69g5kw4ymx%2522%253A%257B%2522id%2522%253A%2522clxuq704500752a69g5kw4ymx%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clxuq704500762a69yoj1jd12%2522%253A%257B%2522id%2522%253A%2522clxuq704500762a69yoj1jd12%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxtg87yn00032a69gisj4210%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clxtg86fr000hd9f89rtlhhys%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clxtg8gk500652a69yv83z3oe%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clxtg87yn00032a69gisj4210%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A19.853249475890976%257D)

App.tsx
```
export default function App() {
  const list = useListData({
    initialItems: [
      {id:1, name: 'Permanent'},
      {id:2, name: 'Temporary'},
    ],
    initialSelectedKeys: ['Permanent'],
    getKey: item => item.name
  });

  return (
    <div className="App">
      <div>
          <JobTypeSelect items={list.items} list={list} className="" />
      </div>
    </div>
  );
}
```

JobTypeSelect.tsx
```
interface Props<T> {
  className: string;
  list: ListData<T>;
  items: { id: number; name: string }[];
}

export const JobTypeSelect = <T,>({ className, list, items }: Props<T>) => {
  return (
    <ItemSelect label="Job Type" items={items} list={list}>  // <------------------------this is where I get the error
      {(item) => <ListBoxItem id={item.name}>{item.name}</ListBoxItem>}
    </ItemSelect>
  );
};
```

ItemSelect.tsx
```
interface Props<T extends object> extends Omit<SelectProps<T>, "children"> {
  label?: string;
  description?: string;
  errorMessage?: string | ((validation: ValidationResult) => string);
  items?: Iterable<T>;
  list: ListData<T>;
  children: React.ReactNode | ((item: T) => React.ReactNode);
}

export function ItemSelect<T extends object>({
  label,
  description,
  errorMessage,
  children,
  items,
  list,
  ...props
}: Props<T>) {
  return (
    <Select {...props}>
      <Label>
        {label}
      </Label>
      <Button>
        <SelectValue />
        <span aria-hidden="true">▼</span>
      </Button>
      {description && <Text slot="description">{description}</Text>}
      <FieldError>{errorMessage}</FieldError>
      <Popover>
        <ListBox
          items={list.items}
          selectedKeys={list.selectedKeys}
          onSelectionChange={list.setSelectedKeys}
        >
          {children}
        </ListBox>
      </Popover>
    </Select>
  );
}
```

**Reply by NickWoodward** [reply]
saw you in codesandbox, thanks for checking :)

**Comment by NickWoodward** [comment]
Sorry for bumping again!

Have just noticed that the `Select` was working in-spite of passing the `selectedKeys` and `onSelectionChange` properties to the `ListBox` component

Just not sure why neither seem to be working? [Demo link](https://codesandbox.io/p/devbox/test-28j3m5?file=%2Fsrc%2FItemSelect.tsx%3A50%2C21)

```
<ListBox
  items={list.items}
  selectedKeys={list.selectedKeys}
  onSelectionChange={(key) => {
    console.log("not called");
    list.setSelectedKeys(key);
  }}
>
```

**Comment by yihuiliao** [comment]
Because the `<ListBox />` is within a `<Select /> `component, you'll want to set onSelectionChange on `<Select /> ` instead of on `<ListBox>`.  I can look into it more but that should at least help you get started

**Reply by NickWoodward** [reply]
> Because the `<ListBox />` is within a `<Select /> `component, you'll want to set onSelectionChange on `<Select /> ` instead of on `<ListBox>`. I can look into it more but that should at least help you get started

Thanks a lot, appreciate it! :)
